### ISSUE TITLE
ListStore actions should be completed in the bg thread

### ISSUE THREAD

**Issue by oguzkocer** [wordpress-mobile/WordPress-FluxC-Android#1201]
While reviewing the post list feature in WPAndroid, I realized that some `ListStore` actions were being taken in the main thread. We need to move the callback functions to the bg thread so that there won't be any disk reads in the main thread. I don't think there are any disk writes or network requests, but it's also something to watch out for.

I've a temporary branch `issue/fix-list-store-threading` for this just to test things out and it looked to be working from a quick test. That might not be the best approach though, so we can simply delete that branch if there is a better one.

**Comment by malinajirka** [comment]
I've checked the branch and I'm still not sure what's the best practice. I have seen a couple of articles which advice not to use GlobalScope. Even the documentation is pretty clear `Application code usually should use application-defined CoroutineScope, using async or launch on the instance of GlobalScope is highly discouraged.`.

I've created a branch `issue/fix-list-store-threading2` 
- I updated the relevant code in PagedListWrapper
- I'm still not sure what to do about the usage in ListStore - It kind of makes sense it's started with a GlobalScope since it's a singleton, right?

**Comment by oguzkocer** [comment]
I only used the `GlobalScope` to check if it fixes the problem for us, or better put, to find out which actions were being taken in main thread. Having said that, I did not (still do not?) know what the correct thing to do is, so I was going to just research it.

I checked the changes and it's looking good. The `PagedListWrapper` approach looks very similar to what we do in WPAndroid with ViewModels and makes total sense to me.

> I'm still not sure what to do about the usage in ListStore - It kind of makes sense to start it with a GlobalScope since it's a singleton, right?

Although it's a singleton, I don't think we should treat it as such. Is there a way to not use `GlobalScope` here? I checked some other stores and they are also using `GlobalScope`, so I guess it'd be fine to use it here as well for now, but if there is a better way, I'd like us to move all stores in that direction.

If you don't have a specific approach in mind for it, I'm totally fine with just using `GlobalScope` here and opening an issue for all Stores to look into this issue separately. That'd be a low priority issue, but at least it'll be documented.

What do you think?

**Comment by malinajirka** [comment]
>Although it's a singleton, I don't think we should treat it as such. Is there a way to not use GlobalScope here? I checked some other stores and they are also using GlobalScope, so I guess it'd be fine to use it here as well for now, but if there is a better way, I'd like us to move all stores in that direction.

I might have used an in-appropriate term. What I meant is the store is not tied to any ui component/state and it basically has an application scope (or user scope at least). 

>If you don't have a specific approach in mind for it, I'm totally fine with just using GlobalScope here and opening an issue for all Stores to look into this issue separately. That'd be a low priority issue, but at least it'll be documented.

Sounds good to me. I'm still not sure what is the best approach. We could create an application scope in dagger and pass it to the Stores, but I'm still missing the major difference between this approach and using GlobalScope.

Submitted an issue https://github.com/wordpress-mobile/WordPress-Android/issues/9527

**Comment by shiki** [comment]
Re: `GlobalScope`. I'm guessing that one of the benefits of using our own scope is we are not susceptive to any other code calling `GlobalScope.cancel()` which cancels all the jobs under that scope. 

In addition to that, we'd probably also want to define our own `CoroutineDispatcher` since the `Dispatchers.Default` has a [limited thread pool](https://medium.com/capital-one-tech/kotlin-coroutines-on-android-things-i-wish-i-knew-at-the-beginning-c2f0b1f16cff).

**Comment by malinajirka** [comment]
Both really good points, thanks @shiki! What I'm still confused about is how this all works when you provide your own context to the GlobalScope (eg. "GlobalScope.launch(customContext)"). 

(Nevertheless, we currently inject `Dispatchers.getDefault()` which has definitely the issue with the limited thread pool.)

**Comment by oguzkocer** [comment]
Thanks for opening the issue @malinajirka, let's continue this discussion in https://github.com/wordpress-mobile/WordPress-Android/issues/9527 since we'll close this one once #1203 is merged.

I still haven't found the time to look into this. If either of you know how best to proceed with this, I'd love to learn about it.

**Comment by oguzkocer** [comment]
Fixed in #1203.
